docs/src/developer/design/user_api.rst

```diff
@@ -36,133 +55,359 @@ functional-style and cmake-based
+properly configured CMake targets.
+
+Test Project
+============
+
+After targets are built the next step is to test that they were built correctly.
+Testing build targets with CMake often requires:
+
+- finding dependencies of the testing framework,
+- building testing targets which consume the project's build targets, and
+- registering the tests with CTest.
+
+It is also worth noting that tests are often built conditionally (e.g., a
+:term:`build system` typically does not build the tests of dependencies built
+during the "Find Dependencies" step). To that end CMake defines the
+``BUILD_TESTING`` variable; when set to ``TRUE`` tests are built, otherwise they
+are not. Proposed CMaize APIs for testing a project:
+
+.. code-block:: CMake
+
+   cmaize_find_or_build_test_dependency(
+       <name>
+       URL <where_on_the_internet_to_download_from>
+       VERSION <the_version_you_want>
+       BUILD_TARGET <target_to_build>
+       FIND_TARGET <target_representing_package>
+       CMAKE_ARGS <configuration_options_to_set>
+   )
+
+   cmaize_add_test_library(
+       <name>
+       SOURCE_DIR <where_the_source_files_are_located>
+       INCLUDE_DIRS <directories_containing_header_files>
+       DEPENDS <dependency0> <dependency1> ...
+   )
+
+   cmaize_add_test_executable(
+       <name>
+       SOURCE_DIR <where_the_source_files_are_located>
+       INCLUDE_DIRS <directories_containing_header_files>
+       DEPENDS <dependency0> <dependency1> ...
+   )
+
+   # This is actually CTest's add_test command
+   add_test(NAME <name> COMMAND)
+
+   # This is a convenience function for the common scenario where the
+   # add_test call simply calls the executable arising from the
+   # cmaize_add_test_executable
+   cmaize_add_test(
+       <name>
+       SOURCE_DIR <where_the_source_files_are_located>
+       INCLUDE_DIRS <directories_containing_header_files>
+       DEPENDS <dependency0> <dependency1> ...
+   )
+
+All of the test functions are thin wrappers around the non-test functions of
+the same name (e.g., ``cmaize_add_test_library`` wraps ``cmaize_add_library``),
+which hide the logic for including the CTest CMake module, and checking that
+``BUILD_TESTING`` is enabled (if it's not the functions are no-ops).
+
+Install Project
+===============
+
+If the tests are successful (or were skipped) it's on to :term:`package`
+installation. Installation typically requires specifying which targets are
+part of the package, generating the packaging files, and then literally
+moving the targets and files to their final location. The main considerations
+for installing are:
+
+- Collecting sufficient information to be able to install the package including:
+  where it goes, which pieces get installed, and what the runtime dependencies
+  are.
+- Installation should be done in a manner which considers the package manager.
+
+The proposed installation :term:`API` is:
+
+.. code-block:: CMake
+
+   cmaize_add_package(<name> NAMESPACE <namespace>)
```

Comment:
Currently, `cmaize_add_package` can also accept a `VERSION` option and `TARGETS` list, at least for the CMakePackageManager version. `VERSION` specifies the version of the package while `TARGETS` specifies the list of targets to actually include in the package. Do you think either of these are something to include in the general API as well?

Reply:
`TARGETS` should eventually be exposed, but I don't really see a need to expose `VERSION`.